### DISCUSSION TITLE
How to run Pyswmm with a different swmm5.dll?

### DISCUSSION THREAD

**Discussion by Alex199449** [pyswmm/pyswmm#347] [Q&A]
I want to use a different SWMM called SWMM-UrbanEVA in PySWMM. 
For doing that I need to build up PySWMM on the different swmm5.dll.
I have the git version of the SWMM-UrbanEVA, but how are the steps to reproduce PySWMM based on the different SWMM.

 - OS:  Windows,

**Comment by bemcdonnell** [comment]
@abhiramm7, can you provide that simple guide you were developing for the workflow?

**Reply by Alex199449** [reply]
I did the steps and everything works quite well i supose, but what should I do in step 7.
I cant find the <the wheel which corresponds to your system>??

**Reply by abhiramm7** [reply]
hey @Alex199449 , can you share the screenshot of the contents of your swmm folder? I also owe you an email, I'll try get back to you asap.

**Reply by abhiramm7** [reply]
@bemcdonnell 

1. Fork/Clone https://github.com/OpenWaterAnalytics/swmm-python 
2. cd into swmm-toolkit directory
3. Add your modified swmm into swmm-solver folder
4. Run python -m pip install -r build-requirements.txt
5. Run python setup.py build
6. Run python setup.py bdist_wheel
7. Run python -m pip install dist/swmm_toolkit/the wheel which corresponds to your system

**Reply by Alex199449** [reply]
![image](https://user-images.githubusercontent.com/104619293/171876467-7b22c8b2-bb9b-45f9-93b0-6a0876ce0756.png)

**Reply by Alex199449** [reply]
While trying to step 5 I got this error.

(base) C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit>python setup.py build
C:\Users\User\anaconda3\lib\distutils\dist.py:274: UserWarning: Unknown distribution option: 'cmake_process_manifest_hook'
  warnings.warn(msg)


--------------------------------------------------------------------------------
-- Trying "Visual Studio 17 2022" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.22000.
CMake Error at CMakeLists.txt:2 (PROJECT):
  Failed to run MSBuild command:

    C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe

  to get the value of VCTargetsPath:

    Microsoft (R)-Build-Engine, Version 17.2.1+52cd2da31 f├╝r .NET Framework
    Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

    Der Buildvorgang wurde am 07.06.2022 18:10:38 gestartet.
    Projekt "C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_cmake_test_compile\build\CMakeFiles\3.21.0\VCTargetsPath.vcxproj" auf Knoten "1" (Standardziele).
    C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(440,5): error MSB8020: Die Buildtools f├╝r Visual Studio 2019 (Plattformtoolset = "v142") wurden nicht gefunden. Installieren Sie zum Erstellen mithilfe der v142-Buildtools die Visual Studio 2019-Buildtools. Alternativ k├Ânnen Sie ein Upgrade auf die aktuellen Visual Studio-Tools durchf├╝hren, indem Sie das Men├╝ "Projekt" ausw├ñhlen oder mit der rechten Maustaste auf die Projektmappe klicken und dann "Projektmappe neu zuweisen" ausw├ñhlen. [C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_cmake_test_compile\build\CMakeFiles\3.21.0\VCTargetsPath.vcxproj]
    Die Erstellung des Projekts "C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_cmake_test_compile\build\CMakeFiles\3.21.0\VCTargetsPath.vcxproj" ist abgeschlossen (Standardziele) -- FEHLER.

    Fehler beim Buildvorgang.

    "C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_cmake_test_compile\build\CMakeFiles\3.21.0\VCTargetsPath.vcxproj" (Standardziel) (1) ->
    (PrepareForBuild Ziel) ->
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(440,5): error MSB8020: Die Buildtools f├╝r Visual Studio 2019 (Plattformtoolset = "v142") wurden nicht gefunden. Installieren Sie zum Erstellen mithilfe der v142-Buildtools die Visual Studio 2019-Buildtools. Alternativ k├Ânnen Sie ein Upgrade auf die aktuellen Visual Studio-Tools durchf├╝hren, indem Sie das Men├╝ "Projekt" ausw├ñhlen oder mit der rechten Maustaste auf die Projektmappe klicken und dann "Projektmappe neu zuweisen" ausw├ñhlen. [C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_cmake_test_compile\build\CMakeFiles\3.21.0\VCTargetsPath.vcxproj]

        0 Warnung(en)
        1 Fehler

    Verstrichene Zeit 00:00:00.96


  Exit code: 1



-- Configuring incomplete, errors occurred!
See also "C:/Users/User/anaconda3/Lib/site-packages/swmm_toolkit/_cmake_test_compile/build/CMakeFiles/CMakeOutput.log".
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Visual Studio 17 2022" generator - failure
--------------------------------------------------------------------------------

********************************************************************************
scikit-build could not get a working generator for your system. Aborting build.

Building windows wheels for Python 3.9 requires Microsoft Visual Studio 2017.
Get it with "Visual Studio 2017":

  https://visualstudio.microsoft.com/vs/

********************************************************************************

**Reply by abhiramm7** [reply]
Interesting, do you have the cmakelists.txt file in swmm code?

**Reply by abhiramm7** [reply]
If possible, can you send the swmm code you are trying use? I'll try to build pyswmm with your swmm on my end and see what's happening.

**Reply by Alex199449** [reply]
That would be great.

https://git.fh-muenster.de/bh642456/swmm-urbaneva/-/tree/dev-urbaneva
I use this swmm.

**Reply by abhiramm7** [reply]
@Alex199449  owa-swmm has a toolkitapi.c file that provides the functionality for pyswmm to interact with swmm solver. Your version of swmm does not have it. You need that for it to work with pyswmm. We might have to hack a solution to get this to work, so let's try this,

1. Clone this https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/tree/master 
2. Replace the files in the https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/tree/master/src/solver with the version of your swmm files. You should copy the contents of your swmm to this folder. In the end, you will have a folder with your version of swmm and files needed for pyswmm.
3. Move them into swmm-python and try building it again.

**Reply by Alex199449** [reply]
Did it work at your end ?
So for the explanation:
I am using anaconda. Do I have to install swmm-python before or build the folder somewhere?
I did before the files into the swmm_toolkit folder where the .dll files are inside ist that correct?

**Reply by abhiramm7** [reply]
@Alex199449 I want to step back for a bit, how do you plan on using pyswmm after we get it to work with your version of swmm? I want to make sure our efforts are in the right direction.

**Reply by Alex199449** [reply]
Ok. I want to use this tool and implement then smart rain barrels which I can control in Real-Time dependend on the soil-moisture of the new lid from the urban-eva.

**Reply by Alex199449** [reply]
@abhiramm7 do you understand?

**Reply by Alex199449** [reply]
Hey @abhiramm7 
It looks like the python setup.py build is running through. 
But I got the error: 
C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\swmm-solver\src\solver\include\toolkit.h(22): fatal error C1083: Datei (Include) kann nicht geöffnet werden: "toolkit_export.h": No such file or directory [C:\Users\User\anaconda3\Lib\site-packages\swmm_toolkit\_skbuild\win-amd64-3.9\cmake-build\src\swmm\tool
kit\solver.vcxproj]

**Reply by Alex199449** [reply]
@abhiramm7 Did it work at your end to implement the SWMMUrbanEVA inot pyswmm?

**Reply by Alex199449** [reply]
@bemcdonnell and @abhiramm7 
I suppose it is not that easy to get it to run in pyswmm right?
I dont want to bother you, but I think if it is not possible I have to do a different task.
Thank you

**Reply by bemcdonnell** [reply]
@Alex199449, i think what Abhi has been trying to guide you to is to insert your code into the OWA-SWMM codebase rather than trying to mesh the OWA-SWMM code into your project.   Start with something that works - OWA-SWMM.  Follow the process with the OWA-SWMM codebase to get OWA-SWMM to compile, then find the differences between the SWMMUrbanEVA code and OWA-SWMM, insert those lines, then try to compile.  It's not a straightforward merge.  I am quite certain it is possible

**Reply by abhiramm7** [reply]
@Alex199449 sorry for not getting back. I am on vacation and haven't had a chance to look through your code yet. SWMM build info is in https://github.com/OpenWaterAnalytics/Stormwater-Management-Model/wiki/Build-System, it should be a good reference for building swmm. It uses cmake.

**Reply by Alex199449** [reply]
@abhiramm7 and @bemcdonnell 
Hey guys it works.
I could build it.
now I finished also the guide.
When I finished the guide I have to install pyswmm right?
So pip install pyswmm?

**Reply by Alex199449** [reply]
@abhiramm7 and @bemcdonnell
It is working complete in PySWMM! Thank you so much.
I have only one Problem with the report-file.
How can I write the result Parameter in the LID-PErformance Summary.
The picture below shows what I mean.
<img width="874" alt="image" src="https://user-images.githubusercontent.com/104619293/178480045-7664ea4f-b2ae-4c90-bd08-a6adc9e2cc96.png">

Thank you

**Reply by abhiramm7** [reply]
@Alex199449 that's awesome! Congratulations 🥳

**Reply by abhiramm7** [reply]
Can you elaborate on what you mean by write? Do you mean updating the LID summary parameters after simulation?

**Reply by Alex199449** [reply]
@abhiramm7 
Thank you that is so nice!!
Yes in the LID-Reoprt File I have all the Data for Evap Intercep, Evap Surface, Evap Soil and TrEvap Soil.
But in the report File in Lid Performance Summary are zeros for the parameters after the simulation. 
How can I put values behind there like in the columns before for ecample Evap Loss?

**Reply by abhiramm7** [reply]
Oh intresting, I dont think there is a way for us to do it easily. I think it is easy to just process that externally rather than updating the repot file

**Reply by bemcdonnell** [reply]
@Alex199449 are you trying to parse the LID-output file after the simulation?  Doesn't it make more sense to get those values using pyswmm before closing?

**Reply by Alex199449** [reply]
Hey @bemcdonnell, 
Sry for my late respond. 
Yes that would be great how can I get all the values from the lids and maybe save them later in a dataframe and then in a csv or something like that?
